### ISSUE TITLE
%S doesn't escape string template characters and/or unable to escape template characters

### ISSUE THREAD

**Issue by alex2069** [square/kotlinpoet#439]
Not sure if this could be considered a bug or not.

For the given function:
```kotlin
val stringFromUnknownSource = "Some string\n" + "$" + "annoyingUser" + " is annoying."
val getStringFunc = FunSpec.builder("getString").addStatement("return %S", stringFromUnknownSource).build()
```

Results in:
```kotlin
fun getString() = """
  |Some string
  |$annoyingUser is annoying.
  """.trimMargin()
```

Which fails to compile since the user or whatever was trying to be a smart ass (or perhaps they really did just make a mistake heh).

I would kind of expect the output of `return %S` to be a pure string and nothing more. Whereas if I wanted a string with side-effects (template string) I would use `return "%L"`? Not sure TBH as I suspect it might be intentional?

So I suppose the question is - where should the responsibility of escaping come into it? The documentation kind of implies it handles the escaping for you:
> `%S` escapes the value as a *string*, wraps it with double quotes, and emits that. For example, `6" sandwich` is emitted `"6\" sandwich"`.

Ironically however, when I tried to escape the `$` by adding a `\` in front of it, the `%S` then escapes that, thwarting my escaping attempts.

As a workaround I've had to replace all `%S` usages with `%L` and a pre-escaped string.
So the above turns into:
```kotlin
fun String.toKString() = "\"\"\"${replace("\$", "\${'\$'}")}\"\"\""
FunSpec.builder("getString").addStatement("return %L", stringFromUnknownSource.toKString()).build()
```

And Generates:
```kotlin
fun getString() = """Some string
${'$'}annoyingUser is annoying."""
```

-----

So just before submitting this I figured I should check for related issues and found #427 and #428 - so not sure about this now. However they are both closed and the consensus was that it didn't seem intuitive to use `%L` for templates (which I disagree with - I expected `%S` to be escaping). 

At the moment it's doing something in between (which is counter intuitive). This also means there are two ways to emit a templated string `%S` and `"%L"`, but no way to have a "pure" string - which I would argue is desired most of the time (also it appears `%L` with a multi-line string is altered by the code indentation/formatting).

If the behaviour remains unchanged then we can never reliably use `%S` from unknown sources (which I'd imagine is frequently the case; from user's annotation declarations or what have you).

**Comment by swankjesse** [comment]
Should we have different escape sequences for strings with interpolation (`%P` for pattern ) vs. those without (`%S` for string) ?

I expect that patterns are going to be less common.

**Comment by Egorand** [comment]
This doesn't seem to solve the problem of formatting strings coming from unknown sources: should you go with `%P` or `%S` for those?

**Comment by swankjesse** [comment]
Definitely %S. You should have to opt-in to executable strings!

**Comment by Egorand** [comment]
@alex2069 We've just merged #473, which makes sure `%S` escapes `$` signs properly. For now this would mean that it's not possible to create template strings with `%S` - you'd have to use `%L`. Would love to hear your thoughts on this.

**Comment by Egorand** [comment]
As for templated strings, in future we might introduce a new placeholder that does proper formatting and doesn't escape `$` symbols. I feel like it's not commonly used though, so consider it low priority.

**Comment by frko** [comment]
After using 1.0.0-SNAPSHOT my previously working string templates got escaped by the patch landed for #473. I fully understand the rationale to escape while using ```%S```. The solution to use ```%L``` feels a bit cumbersome as I need to start escaping strings in my ```KP``` templates to get non escaped string values. I am not sure if the proposed ```%P``` would entail escaping wise but for me at this point but to keep ```KP``` templated strings clean a ```%LS``` meaning ```literal string``` might suffice for the most common usecase.

**Comment by alex2069** [comment]
> @alex2069 We've just merged #473, which makes sure `%S` escapes `$` signs properly. For now this would mean that it's not possible to create template strings with `%S` - you'd have to use `%L`. Would love to hear your thoughts on this.

I don't really have many thoughts on the matter anymore. I've been using my `%V` #481 branch with jitpack and intend to shadowjar it so it wont break other peoples normal KotlinPoet versions - I often don't know the type that I'm trying to emit so TBH I doubt I'll go back to normal versions (doing so would require a ton of if/when checks).

**Comment by Egorand** [comment]
Closing this based on #473, and we'll follow up at some point with a better solution for template strings.
